Update for Qt6

jmlich opened the pull request:
Even if it is not finished, I would like to open a discussion about what changes are needed for inclusion in the master. I may open separate pull requests for every part.

jmlich commented:
Seems to be working as expected now

Thaodan commented:
Don't include merge commits. 
Some of the commits could be smashed as for example those that address Qt6 support.
Also remove commits that drop changes from earlier commits.

Thaodan commented on src/micuconversions.cpp:
```diff
@@ -326,7 +330,11 @@ QString MIcuConversions::icuDatePatternEscaped(const QString &str)
     QString value;
+#if QT_VERSION < 0x051500
     QRegExp regexp("^[^@]+@.*"+QRegExp::escape(option)+"=([^@=;]+)($|;.*$)");
+#else
```
QRegularExpressions exist since Qt 5.0, unless you using a Qt 5.15 feature they should work earlier.

jmlich replied:
QRegExp was removed

jmlich committed:
[mlocale] Fix build with Qt6 JB#52793

qt5 in .pro replaced by qt$${QT_MAJOR_VERSION}
QTextCodec is deprecated in Qt6
QIODevice::WriteOnly was moved to QIODeviceBase::WriteOnly
QRegExp was replaced by QRegularExpression
QString::SkipEmptyParts was moved to Qt::SkipEmptyParts

Co-authored-by: Chupligin Sergey <neochapay@gmail.com>

Thaodan reviewed (changes requested):
Please change the commit message style similar to:
https://github.com/sailfishos/qt-mobility-haptics-ffmemless/pull/1#pullrequestreview-1598788046

Rest looks good to me.

pvuorela commented on src/mlocale.cpp:
```diff
@@ -4770,8 +4803,9 @@ QString MLocalePrivate::formatPhoneNumber( const QString& phoneNumber,
   // first do sanity check of the input string
-  QRegExp rx( "\\+?\\d*" );
-  if ( ! rx.exactMatch( phoneNumber ) )
+  QRegularExpression rx( QRegularExpression::anchoredPattern("\\+?\\d*") );
```
anchoredPattern() introduced in qt 5.15. We still need support for older versions too.

pvuorela commented on src/mlocale.h:
```diff
@@ -28,7 +28,9 @@
 
 class QString;
-class QStringList;
+#if QT_VERSION < 0x051500
```
Why this? All versions have QStringList in the API and even if those didn't, the class declaration shouldn't have too much side-effects?

pvuorela commented on src/mdebug.h:
```diff
@@ -65,7 +65,11 @@ class MDebug
     QtMsgType convertMsgType(int type);
 
     struct Stream {
+#if QT_VERSION < 0x051500
```
QIODeviceBase is a Qt6 thing?

pvuorela commented on src/mcharsetdetector.cpp:
```diff
@@ -500,6 +502,12 @@ QList<MCharsetMatch> MCharsetDetector::detectAll()
 QString MCharsetDetector::text(const MCharsetMatch &charsetMatch)
 {
     Q_D(MCharsetDetector);
+    #if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
```
Elsewhere the #ifs not indented.

pvuorela commented on src/mcharsetdetector.cpp:
```diff
@@ -500,6 +502,12 @@ QList<MCharsetMatch> MCharsetDetector::detectAll()
     Q_D(MCharsetDetector);
+    #if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
+    if(charsetMatch.name().toLatin1().isEmpty() || charsetMatch.name().toLatin1() != "UTF-8") {
+        qWarning() << "Don`t use non utf8 charset";
```
Wouldn't it make the class detecting encodings a bit pointless if it can be used only with utf8? Could use core5compat and use the QTextCodec?

pvuorela commented on tests/ut_translations/ut_translations.cpp:
```diff
@@ -155,12 +155,15 @@ void Ut_Translations::testOriginalQtTr()
     // Everything should be untranslated now:
     QCOMPARE(tr(qPrintable(string_id)), string_id);
 
+#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
```
Why these variations?

pvuorela commented:
Replaced by #7
